File: src/07_Advanced_Pre_DSA/Templates_Generics.cpp
/**
 * Level 7: Advanced Pre-DSA & Standard Libraries
 * File: Templates_Generics.cpp
 * Topic: Template classes and functions, Monomorphization, template specialization
 */

#include <iostream>

// Generic Template Function
template <typename T>
T getMin(T a, T b) {
    return (a < b) ? a : b;
}

// Generic Template Pair Class (Essential for Graph edges and DP states!)
template <typename K, typename V>
class MyPair {
public:
    K first;
    V second;
    MyPair(K f, V s) : first(f), second(s) {}
};

int main() {
    std::cout << "Min int:    " << getMin<int>(10, 20) << "\n";
    std::cout << "Min double: " << getMin<double>(3.14, 2.71) << "\n";

    MyPair<int, std::string> p(1, "Vertex_A");
    std::cout << "Pair: (" << p.first << ", " << p.second << ")\n";

    return 0;
}

File: src/07_Advanced_Pre_DSA/Templates_Generics.java
/**
 * Level 7: Advanced Pre-DSA & Standard Libraries
 * File: Templates_Generics.java
 * Topic: Generics <T>, Type Erasure, Bounded type parameters
 */

public class Templates_Generics {
    // Generic method with bounded type parameter
    public static <T extends Comparable<T>> T getMin(T a, T b) {
        return (a.compareTo(b) < 0) ? a : b;
    }

    // Generic Pair Class
    static class MyPair<K, V> {
        K first;
        V second;
        MyPair(K f, V s) { this.first = f; this.second = s; }
    }

    public static void main(String[] args) {
        System.out.println("Min int: " + getMin(10, 20));
        System.out.println("Min double: " + getMin(3.14, 2.71));

        MyPair<Integer, String> p = new MyPair<>(1, "Vertex_A");
        System.out.println("Pair: (" + p.first + ", " + p.second + ")");
    }
}
